### PR TITLE
Gitlab integration

### PR DESCRIPTION
**Pull request by wandersoncferreira**
Supported in this PR so far:

- [x] [REQUIRED] `code-review-core-pullreq-infos` - Used in rendering
- [x] [REQUIRED] `code-review-code-pullreq-diff`   - Used in rendering
- [ ] [v2] `code-review-code-commit-diff`
- [ ] [v2] `code-review-commit-comments`
- [x] [REQUIRED] `code-review-core-send-review`
- [x] [REQUIRED] `code-review-core-send-replies`
- [ ] [v2] `code-review-core-get-labels`
- [ ] [v2] `code-review-core-set-labels`
- [ ] [v2] `code-review-core-get-assignees`
- [ ] [v2] `code-review-core-set-assignee`
- [ ] [v2] `code-review-core-get-milestones`
- [ ] [v2] `code-review-core-set-milestone`
- [ ] [v2] `code-review-core-set-title`
- [ ] [v2] `code-review-core-set-description`
- [ ] [v3] `code-review-core-merge`
- [ ] [Not sure if possible to support] `code-review-core-set-reaction`
- [ ] [Not sure if possible to support] `code-review-core-delete-reaction`

The goal is to finish all REQUIRED methods for the main workflow of the package.

#### Success

- https://gitlab.com/inkscape/inkscape/-/merge_requests/3665
- https://gitlab.com/code-review-experiment/default/-/merge_requests/1
- https://gitlab.com/inkscape/inkscape/-/merge_requests/3628
- https://gitlab.com/gitlab-org/gitlab-runner/-/merge_requests/3220
- https://gitlab.com/inkscape/inkscape/-/merge_requests/3534

### CODE REVIEW

**Comment by ryanprior** [comment]
I've been reading the code and made a few small changes on my fork. Feel free to pull into your branch from https://github.com/ryanprior/code-review/tree/feat/gitlab-integration

**Comment by wandersoncferreira** [comment]
@ryanprior you mentioned to have looked at Gitlab's APIs in the past, this Graphql query below is the best I could do to get comments related to a PR:

```elisp
query {
  repository:project(fullPath: "code-review-experiment/default") {
    pullRequest:mergeRequest(iid: "1") {
      id
      comments:notes(first: 50){
        nodes {
          databaseId:id
          bodyText: body
          author {
            login:username
          }
          createdAt
          updatedAt
          system
          position {
            height
            newLine
            newPath
            oldLine
            oldPath
            width
            x
            y
          }
         }
      }
    }
  }
}
```

But I could not get an explicit way to say correlate a comment to a reply... only using the `position.oldLine` and `position.oldPath` values. Seems to be a good first approximation IMHO.

File: code-review-gitlab.el [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+                        :auth 'code-review
+                        :host code-review-gitlab-host
+                        :errorback #'code-review-gitlab-errback
+                        :callback (lambda (&rest _)))))
+         (oref replies replies)))
+
+      (deferred:nextc it
+        (lambda (_x)
+          ;; seems like we need to wait a bit for gitlab's API to update the new reply record
+          (sit-for 0.5)
```
Comment (tibonarium) [inline code-review-gitlab.el:289]: Maybe use `deferred:wait`

**Comment by wandersoncferreira** [comment]
In order to finish the `code-review-core-send-review` function and have a bare minimum working version we need two things:

- [ ] compute the *line number* of a particular diff position (their API requires the line number -.-")
- [ ] identify if the comment was made in a Added/Removed/Unchanged line

These informations should be added to the `code-review-local-comment-section` object.

File: code-review-gitlab.el (hunk 1 above)
Comment (wandersoncferreira) [reply code-review-gitlab.el:290]: Good idea! thanks @tibonarium

**Comment by wandersoncferreira** [comment]
Update: the integration requires more testing now and some DRY up in the functions around the feature. After finishing this bit, is clear to me that intgrations with other forges are not so smooth 👎 

I created an issue to track some better code structuring to allow better maintainability of this package.

Anyway, if someone is a heavy user of Gitlab and could pull this branch and test the basic workflow would be great:

1. Fetch the PR
2. Add comments
3. Reply to comments
4. Send review

I'm not familiar enough with gitlab but seems like you can't "request for changes" there.. there is only one option as "Approve" a PR afaict.

File: code-review.el [hunk 2]
```diff
@@ -199,15 +200,48 @@ If you want to provide a MSG for the end of the process."
       (deferred:$
         (deferred:parallel
           (lambda () (code-review-core-diff-deferred obj))
-          (lambda () (code-review-core-infos-deferred obj)))
+          (lambda ()  (code-review-core-infos-deferred obj)))
```
Comment (wandersoncferreira) [inline code-review.el:203]: Remove unnecessary whitespace

File: code-review-gitlab.el [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+(defcustom code-review-gitlab-graphql-host "gitlab.com/api"
+  "Host for Graphql in Gitlab."
+  :group 'code-review-gitlab
+  :type 'string)
+
+(defclass code-review-gitlab-repo (code-review-db-pullreq)
+  ((callback :initform nil)))
+
+(defvar code-review-log-file)
+(defvar code-review-gitlab-line-diff-mapping nil)
```
Comment (wandersoncferreira) [inline code-review-gitlab.el:48]: Add documentation

File: code-review-gitlab.el [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+
+(defun code-review--log (origin msg)
```
Comment (wandersoncferreira) [inline code-review-gitlab.el:50]: Move this function to utils file.. same used in code-review-github.el

File: code-review-gitlab.el [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+  (let-alist m
+    (code-review--log
+     "code-review-gitlab-errback"
+     (prin1-to-string m))
+    (message "Unknown error talking to Gitlab: %s" m)))
+
+(defun code-review-gitlab-fix-diff (pr-changes)
+  "Get all PR-CHANGES and produce standard git diff."
+  (let ((d
+         (-map
```
Comment (wandersoncferreira) [inline code-review-gitlab.el:76]: use `-reduce`

File: code-review-gitlab.el [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+                                  (a-get c 'resolvable)))
+                           raw-comments))
+         (grouped-comments (-group-by
+                            (lambda (c)
+                              (let ((line (or (a-get-in c (list 'position 'oldLine))
+                                              (a-get-in c (list 'position 'newLine))))
+                                    (path (a-get-in c (list 'position 'oldPath))))
+                                (concat path ":" (number-to-string line))))
+                            review-comments))
+         (comment->code-review-comment (lambda (c)
```
Comment (wandersoncferreira) [inline code-review-gitlab.el:237]: Add fn in newline

**Review by wandersoncferreira — COMMENTED** [review]
Some comments for improvements.

Side note:

1. While reviewing this PR on `code-review` sometimes my cursor position was not
placed correctly, where the comment block was added, it needs further
investigation.

2. I am really annoyed by not being able to use "fullscreen" to review my PR. A
   change to choose how better handle window splitting is needed.
